geanylua/glspi_doc.c

```diff
@@ -92,34 +83,51 @@ static const gchar* doc_idx_to_filename(gint idx) {
 
 
 
-/* Actvate and focus the specified document */
+/* Switch to the specified document tab */
 static gint glspi_activate(lua_State* L)
 {
-	gint idx=-1;
-	if (lua_gettop(L)>0) {
-		if (lua_isnumber(L,1)) {
-			idx=(lua_tonumber(L,1));
-			if (idx<0) { /* Negative number refers to (absolute) GtkNotebook index */
-				idx=(0-idx)-1;
-				if (idx>=gtk_notebook_get_n_pages(NOTEBOOK)) { idx=-1;}
-			} else { /* A positive number refers to the geany->documents_array index */
-				idx=doc_idx_to_tab_idx(idx-1);
+	gint lua_idx = -1;
```

Comment:
The scope of `lua_idx` could be reduced, and it probably would make things a bit more clear (is that used for anything else, and what is that `-1` about while `-1` is supposed to mean "first tab"?)

Reply:
I wouldn't mind reducing scope, but I've been told changing previously established functionality should be avoided because it's unknown how many people might depend on the old behavior.

Indexing of notebook and documents array don't necessarily match, so might have to split into two separate or create a new lua function to convert between document and notebook indices.

Regarding the calculation...

-1 ⇒ first tab ⇒ 0
0 - (-1) = 1
1-1 = 0

Reply:
I think `lua_idx` is initialized to avoid compiler warnings / errors.  Don't think the particular value matters.

`tab_idx` has to be initialized to a negative number to ensure correct return value.

Reply:
By reducing scope I just mean moving the variable declaration inside the `if (lua_isnumber(L, 1)) {}` block where it's used.

As for the initialization, the compiler probably wouldn't complain as the initial value is never used, but I get that some people like initializing everything. I just think that `-1` is a poor choice here as IIUC `-1` means something specific in the code below, while `0` would have been better as it's the only clearly invalid value for that same code.

On the end it wouldn't change anything on how this behaves, it would just make the implementation a hair clearer (IMO).

Reply:
> But since it looks like this function never behaved as intended,

It probably worked in the past but for the return value: part of the problematic code was likely "fine" before the internals of Geany's editor tabs was modified a few years ago.

Reply:
That type of scope... I'll change it.  I probably just left the variable declaration where it was when I split the variable.  A lot of code in geany seems to declare variables at the beginning of functions.